app/src/main/java/com/google/android/gms/samples/wallet/CheckoutActivity.kt

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.android.gms.samples.wallet
+
+import android.app.Activity
+import android.app.AlertDialog
+import android.content.Intent
+import android.os.Bundle
+import android.util.Log
+import android.view.View
+import android.widget.ImageView
+import android.widget.TextView
+import android.widget.Toast
+import com.google.android.gms.common.api.ApiException
+import com.google.android.gms.wallet.*
+import kotlinx.android.synthetic.main.activity_checkout.*
+import org.json.JSONException
+import org.json.JSONObject
+
+/**
+ * Checkout implementation for the app
+ */
+class CheckoutActivity : Activity() {
+    /**
+     * A client for interacting with the Google Pay API.
+     *
+     * @see [PaymentsClient](https://developers.google.com/android/reference/com/google/android/gms/wallet/PaymentsClient)
+     */
+    private lateinit var mPaymentsClient: PaymentsClient
+
+    private val mBikeItem = ItemInfo("Simple Bike", (300 * 1000000).toLong(), R.drawable.bike)
+    private val mShippingCost = (90 * 1000000).toLong()
+    /**
+     * Initialize the Google Pay API on creation of the activity
+     *
+     * @see Activity.onCreate
+     */
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_checkout)
+
+        // Set up the mock information for our item in the UI.
+        initItemUI()
+
+        // Initialize a Google Pay API client for an environment suitable for testing.
+        // It's recommended to create the PaymentsClient object inside of the onCreate method.
+        mPaymentsClient = PaymentsUtil.createPaymentsClient(this)
+        possiblyShowGooglePayButton()
+
+        mGooglePayButton.setOnClickListener { requestPayment() }
+    }
+
+    /**
+     * Determine the viewer's ability to pay with a payment method supported by your app and display a
+     * Google Pay payment button.
+     *
+     * @see [](https://developers.google.com/android/reference/com/google/android/gms/wallet/PaymentsClient.html.isReadyToPay
+    ) */
+    private fun possiblyShowGooglePayButton() {
+        val isReadyToPayJson = PaymentsUtil.isReadyToPayRequest()
+        if (!isReadyToPayJson.isPresent) {
+            return
+        }
+        val request = IsReadyToPayRequest.fromJson(isReadyToPayJson.get().toString()) ?: return
```

Comment:
This is probably coming from the Java version but parsing a JSON in the main thread is a big nope

Reply:
At least I would add a comment recommending not to do it this way.